cilium/cmd/kernel_check.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"text/tabwriter"
+
+	"github.com/cilium/cilium/pkg/datapath/linux/probes"
+	"github.com/spf13/cobra"
+)
+
+var kernelCheckCmd = &cobra.Command{
+	Use:   "kernel-check",
+	Short: "Checks whether machine is supporting required kernel capabilities",
+	Run: func(cmd *cobra.Command, args []string) {
+		kernelCheck()
+	},
+}
+
+func kernelCheck() {
+	probeManager := probes.NewProbeManager()
+	w := tabwriter.NewWriter(os.Stdout, 2, 0, 3, ' ', 0)
+
+	requiredParams := probeManager.GetRequiredConfig()
+	allRequiredParamsEnabled := false
+	for k, v := range requiredParams {
+		if v == false {
+			fmt.Fprintf(w, "FEATURE:\t%s\t SUPPORTED:%t\t REQUIRED: true\t\n", k, v)
+		} else {
+			fmt.Fprintf(w, "FEATURE:\t%s\t SUPPORTED:%t\t REQUIRED: true\t\n", k, v)
+			allRequiredParamsEnabled = true
+		}
+	}
+
+	optionalParams := probeManager.GetOptionalConfig()
+	allOptionalParamsEnabled := false
+	for k, v := range optionalParams {
+		if v == false {
+			fmt.Fprintf(w, "FEATURE:\t%s\t SUPPORTED:%t\t REQUIRED: false\t\n", k, v)
+		} else {
+			fmt.Fprintf(w, "FEATURE:\t%s\t SUPPORTED:%t\t REQUIRED: false\t\n", k, v)
+			allOptionalParamsEnabled = true
+		}
+	}
+	if allRequiredParamsEnabled && allOptionalParamsEnabled == false {
+		fmt.Println("Unable to detect BPF support. Is the kernel config available?")
```

Comment:
So if I understand correctly you consider that we didn't find the config file if we don't have at least one of the required parameters _and_ at least one of the optional ones enabled (Or am I missing something)? If so, this seems a bit arbitrary. Why at least two of them? And what if all BPF options are deactivated in config file?

Instead, maybe we could simply check, here or in `GetRequiredConf()`, if `config.ConfigBpf` is set (to either `"y"` or `"n"`). `CONFIG_BPF` does not have any dependency, so I think it should always be in the config file since its introduction in Linux 3.18 (older than the requirements for Cilium), so it looks sufficiently reliable? What do you think?

Reply:
> So if I understand correctly you consider that we didn't find the config file if we don't have at least one of the required parameters and at least one of the optional ones enabled (Or am I missing something)

@qmonnet I'm looping through all required ond optional parameters (keeping track of them) and only if all required as well as all optional parameters are completely set to `false`, I'm printing the error.

Reply:
I might be missing something, but I read this:
```
if allRequiredParamsEnabled && allOptionalParamsEnabled == false {
```
as: “if we found at least one required parameter enabled (i.e. `allRequiredParamsEnabled is true`), and there is no optional parameter enabled at all (`allOptionalParamsEnabled == false`), then print message about missing config file”. Am I misreading it?

Even if I read it wrong and if it does print the message if all required and optional parameters are set to `false`, it does not tell you that the config file is missing: We might well have read that config file and found all parameters at false, couldn't we?

Reply:
@qmonnet thanks, you're right of course. Actually I wanted to keep track of of the required and optional params and only if _all_ of them are set to `false`, print the error message. But the implementation is of course not following that approach, sorry 🤦‍♂️

Reply:
>  Am I misreading it?

@qmonnet no, no, it's all my bad 🙁

Reply:
> [...]  check, here or in GetRequiredConf(), if config.ConfigBpf is set (to either "y" or "n").  so it looks sufficiently reliable? What do you think?

@qmonnet sounds like a clean solution, what do you think @joestringer